sql/migrate/migrate.go

```diff
@@ -127,303 +124,144 @@ func Schema(s *schema.Schema) StateReader {
 }
 
 type (
-	// Dir represents a versioned migration directory.
-	Dir struct {
-		fs        fs.FS
-		conn      Driver
-		pattern   string
-		templates []struct{ N, T *template.Template }
-	}
+	// FS describes the methods needed to work with the Planner.
+	FS interface {
+		// List all present migration files. Usually *.sql files.
+		List() ([]string, error)
 
-	// DirOption allows configuring the Dir
-	// using functional options.
-	DirOption func(*Dir) error
-)
+		// Read the contents of a file by name.
+		Read(name string) ([]byte, error)
 
-// NewDir creates a new workspace directory based
-// on the given configuration options.
-func NewDir(opts ...DirOption) (*Dir, error) {
-	d := &Dir{
-		fs: dirFS{
-			dir: "migrations",
-			FS:  os.DirFS("migrations"),
-		},
-		pattern: "*.sql",
-	}
-	for _, opt := range opts {
-		if err := opt(d); err != nil {
-			return nil, err
-		}
-	}
-	if len(d.templates) == 0 {
-		d.templates = []struct{ N, T *template.Template }{defaultTemplate}
+		// Write the contents to the file by name.
+		Write(name string, data []byte, perm fs.FileMode) error
+
+		// Remove a file by name.
+		Remove(name string) error
 	}
-	return d, nil
-}
 
-type (
-	// dirFS wraps the os.DirFS with additional writing capabilities.
-	dirFS struct {
-		fs.FS
-		dir string
+	// Printer wraps the methods for naming and dumping a migration plan to one or more files.
+	Printer interface {
+		// Print prints the given Plan.
+		// The first return argument contains a slice of filenames.
+		// The second return argument is meant to hold the contents for each filename.
+		// The length of the filenames-slice and contents-slice must be equal.
+		Print(*Plan) ([]string, [][]byte, error)
```

Comment:
Any particular reason not to use a map?

Reply:
<img src="https://media4.giphy.com/media/ggGhuVtIt41J3oY6Gp/giphy.gif"/>